### PR TITLE
[Bugfix] [DSA] [Hisparse] Broadcast TP Rank 0 Topk Indexes to other TPs

### PR DESCRIPTION
**Pull request by xz-keg**
## Motivation

The current implementation of topk.cu is not stable(Given the same input, always output the same output). For DSA models, when many indexes score the same and rank around 2048,  different TPs may select different Top-k. (See test in #24819 )

This bug may only reduce accuracy by a little bit as the selected alternative top-k s usually have very low attention weight due to their ranks being around 2048.
But it is vital for Hisparse implementation and further strategy development, as the index selection also affects the buffer on different GPUs.  @hzh0425 

## Modifications

Here I add an environment param SGLANG_DSA_TOPK_BROADCAST  that changes the computation flow to broadcast the Topk indexes of rank 0 to other TP ranks. 

When SGLANG_DSA_TOPK_BROADCAST is set to False, the original implementation is kept. 

I set default SGLANG_DSA_TOPK_BROADCAST to False for better merge. Currently the broadcast version is a little slower(~3%) than the inaccurate baseline due to broadcast cost. 

To run in broadcast mode, run 
``SGLANG_DSA_TOPK_BROADCAST=true python -m sglang.launch_server ...``

## Accuracy Tests

GLM-5.1-FP8

GSM 8K(5-shot, full set 1314 samples, default config) :

Baseline: 96.57
Broadcast: 96.80

## Speed Tests and Profiling

Serve command: 

8x H200, TP=8, GLM-5.1-FP8

>python -W ignore -m sglang.launch_server \
    --model-path $MODEL_PATH \
    --tp 8 \
    --trust-remote-code  --load-format auto \
    --cuda-graph-bs 1 2 4 8 16 32 64 \
    --max-running-requests 64 \
    --mem-fraction-static 0.85 \
    --host 0.0.0.0  --port 30000 \
    --disable-radix-cache \
     --enable-hisparse --hisparse-config '{"top_k":2048,"device_buffer_size":8192,"host_to_device_ratio":2}'

Test command (8k input 1k output, concurrency 64)

>python -m sglang.bench_serving \
  --backend sglang \
  --host 127.0.0.1 \
  --port 30000 \
  --dataset-name random-ids \
  --random-input-len 8192 \
  --random-output-len  1024\
  --num-prompts 1000 \
  --max-concurrency 64 \
  --output-file sglang_random.jsonl \
  --output-details

Baseline: 

>============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Max request concurrency:                 64
Successful requests:                     1000
Benchmark duration (s):                  845.00
Total input tokens:                      4077386
Total input text tokens:                 4077386
Total generated tokens:                  510855
Total generated tokens (retokenized):    510116
Request throughput (req/s):              1.18
Input token throughput (tok/s):          4825.31
Output token throughput (tok/s):         604.56
Peak output token throughput (tok/s):    1515.00
Peak concurrent requests:                69
Total token throughput (tok/s):          5429.88
Concurrency:                             62.65
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   52940.03
Median E2E Latency (ms):                 51056.81
P90 E2E Latency (ms):                    96141.15
P99 E2E Latency (ms):                    110390.02
---------------Time to First Token----------------
Mean TTFT (ms):                          3209.07
Median TTFT (ms):                        1321.63
P99 TTFT (ms):                           28330.45
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          100.77
Median TPOT (ms):                        98.01
P99 TPOT (ms):                           219.30
---------------Inter-Token Latency----------------
Mean ITL (ms):                           97.62
Median ITL (ms):                         43.50
P95 ITL (ms):                            492.88
P99 ITL (ms):                            732.43
Max ITL (ms):                            26201.88
==================================================

Broadcast:
>============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Max request concurrency:                 64
Successful requests:                     1000
Benchmark duration (s):                  884.23
Total input tokens:                      4077386
Total input text tokens:                 4077386
Total generated tokens:                  510855
Total generated tokens (retokenized):    509966
Request throughput (req/s):              1.13
Input token throughput (tok/s):          4611.24
Output token throughput (tok/s):         577.74
Peak output token throughput (tok/s):    1411.00
Peak concurrent requests:                68
Total token throughput (tok/s):          5188.99
Concurrency:                             62.61
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   55364.68
Median E2E Latency (ms):                 53651.77
P90 E2E Latency (ms):                    100437.73
P99 E2E Latency (ms):                    113721.96
---------------Time to First Token----------------
Mean TTFT (ms):                          3181.58
Median TTFT (ms):                        1368.61
P99 TTFT (ms):                           26087.60
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          105.81
Median TPOT (ms):                        103.59
P99 TPOT (ms):                           212.36
---------------Inter-Token Latency----------------
Mean ITL (ms):                           102.43
Median ITL (ms):                         45.82
P95 ITL (ms):                            522.31
P99 ITL (ms):                            741.28
Max ITL (ms):                            24069.70
==================================================

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #26616204562](https://github.com/sgl-project/sglang/actions/runs/26616204562)
Latest PR Test (Extra): :x: [Run #26616204432](https://github.com/sgl-project/sglang/actions/runs/26616204432)

### CODE REVIEW

**Comment by xz-keg** [comment]
I've added a speed and accuracy test result for this.

**Comment by xz-keg** [comment]
documentation for SGLANG_DSA_TOPK_BROADCAST is also added

**Comment by xiezhq-hermann** [comment]
Nice catch, and this seems like a reasonable fix for top-k selection. @DarkSharpness, could you take a look and leave a comment?

Also, could you help me understand why this is a bigger issue with HiSparse? Host memory retains the full-sequence KV cache regardless, so I would expect HiSparse to behave the same as it does without HiSparse.

**Comment by xz-keg** [comment]
> Nice catch, and this seems like a reasonable fix for top-k selection. @DarkSharpness, could you take a look and leave a comment?
> 
> Also, could you help me understand why this is a bigger issue with HiSparse? Host memory retains the full-sequence KV cache regardless, so I would expect HiSparse to behave the same as it does without HiSparse.

Thanks for the comment. 

For normal inference, this mismatch do not influence result a lot because the mismatch part of the original and alternative indexes have low attention scores so the result is almost the same after softmax attention(10% difference->~0.1% difference after attention).  It may only affect the  accuracy of RL training. 

For Hisparse, you're right that it may not influence the current workflow by a lot since everything has a CPU-backup. 

But I think it is a must-fix bug before any further strategy development because this mismatch influence severely what is on the kv buffer(which is distinct and do not pass a softmax so 10% difference is 10% difference and will influence further if some new method requires buffer alignment across TPs).  

Actually I'm planning to develop a disk-based multi-level Hisparse because CPU memory is usually limited too(and they are becoming more and more expensive). In this case this bug becomes a severe obstacle.

**Comment by DarkSharpness** [comment]
@xiezhq-hermann @xz-keg Yes. Current topk algorithm is not very stable (especially in RL, which might be fatal). Broadcast from TP=0 might be a quick workaround (though NCCL broadcast is not very efficient in decoding).

In the future we need to implement some more stable variant of topk.

**Comment by xz-keg** [comment]
> @xiezhq-hermann @xz-keg Yes. Current topk algorithm is not very stable (especially in RL, which might be fatal). Broadcast from TP=0 might be a quick workaround (though NCCL broadcast is not very efficient in decoding).
> 
> In the future we need to implement some more stable variant of topk.

As for stable topk, #24819 offers tests to test such algorithm. In fact we fallback to torch.topk during our RL (but with high cost because it is not as efficient as topk.cu).  

Maybe let each TP rank computes part of the topk and broadcast to each other is better? This may be a future direction too.

**Comment by DarkSharpness** [comment]
> > @xiezhq-hermann @xz-keg Yes. Current topk algorithm is not very stable (especially in RL, which might be fatal). Broadcast from TP=0 might be a quick workaround (though NCCL broadcast is not very efficient in decoding).
> > In the future we need to implement some more stable variant of topk.
> 
> As for stable topk, #24819 offers tests to test such algorithm. In fact we fallback to torch.topk during our RL (but with high cost because it is not as efficient as topk.cu).
> 
> Maybe let each TP rank computes part of the topk and broadcast to each other is better? This may be a future direction too.

I've thought about this possibility (partial topk + broadcast) before. The problem is that I'm not sure whether the page allocation on each rank is deterministic (FYI, we fused the page table transform in topk for normal code path). Different page allocation on different rank may lead to fatal error in this case. Also, the communication cost is non-trivial here (topk kernel can reach up to 30% HBM memory bandwidth, far exceeding the NVLink bandwidth)

**Comment by xz-keg** [comment]
> I've thought about this possibility (partial topk + broadcast) before. The problem is that I'm not sure whether the page allocation on each rank is deterministic (FYI, we fused the page table transform in topk for normal code path). Different page allocation on different rank may lead to fatal error in this case. Also, the communication cost is non-trivial here (topk kernel can reach up to 30% HBM memory bandwidth, far exceeding the NVLink bandwidth)

To be clear, the broadcast is about top-k indexes, not the full top-k tensors. Top-k tensors are still from device_buffer+swap-in kernel. 

Only the indexes are broadcasted, so 2048(1024/512)*4 bytes(int32) for each sample.

**Comment by xz-keg** [comment]
Since the path changes from nsa.nsa_indexer to a new file named dsa.dsa_indexer, I'll edit accordingly.

**Review by DarkSharpness — APPROVED** [review]
LGTM. Could you please resolve the conflict?

**Comment by xz-keg** [comment]
> LGTM. Could you please resolve the conflict?

conflict resolved and speed test benchmark updated
